homeassistant/components/bom/sensor.py

```diff
@@ -0,0 +1,357 @@
+        }
+
+        return attr
+
+    @property
+    def unit_of_measurement(self):
+        """Return the units of measurement."""
+        return SENSOR_TYPES[self._condition][1]
+
+    def update(self):
+        """Update current conditions."""
+        self.bom_data.update()
+
+
+class BOMCurrentData:
+    """Get data from BOM."""
+
+    def __init__(self, station_id):
+        """Initialize the data object."""
+        self._zone_id, self._wmo_id = station_id.split(".")
+        self._data = None
+        self.last_updated = None
+
+    def _build_url(self):
+        """Build the URL for the requests."""
+        url = (
+            f"http://www.bom.gov.au/fwo/{self._zone_id}"
+            f"/{self._zone_id}.{self._wmo_id}.json"
+        )
+        _LOGGER.debug("BOM URL: %s", url)
+        return url
+
+    @property
+    def latest_data(self):
+        """Return the latest data object."""
+        if self._data:
+            return self._data[0]
+        return None
+
+    def get_reading(self, condition):
+        """Return the value for the given condition.
+
+        BOM weather publishes condition readings for weather (and a few other
+        conditions) at intervals throughout the day. To avoid a `-` value in
+        the frontend for these conditions, we traverse the historical data
+        for the latest value that is not `-`.
+
+        Iterators are used in this method to avoid iterating needlessly
+        through the entire BOM provided dataset.
+        """
+        condition_readings = (entry[condition] for entry in self._data)
+        reading = next((x for x in condition_readings if x != "-"), None)
+
+        if isinstance(reading, (int, float)):
+            return round(reading, 2)
+        return reading
+
+    def should_update(self):
+        """Determine whether an update should occur.
+
+        BOM provides updated data every 30 minutes. We manually define
+        refreshing logic here rather than a throttle to keep updates
+        in lock-step with BOM.
+
+        If 35 minutes has passed since the last BOM data update, then
+        an update should be done.
+        """
+        if self.last_updated is None:
+            # Never updated before, therefore an update should occur.
+            return True
+
+        now = dt_util.utcnow()
+        update_due_at = self.last_updated + datetime.timedelta(minutes=35)
+        return now > update_due_at
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Get the latest data from BOM."""
+        if not self.should_update():
+            _LOGGER.debug(
+                "BOM was updated %s minutes ago, skipping update as"
+                " < 35 minutes, Now: %s, LastUpdate: %s",
+                (dt_util.utcnow() - self.last_updated),
+                dt_util.utcnow(),
+                self.last_updated,
+            )
+            return
+
+        try:
+            result = requests.get(self._build_url(), timeout=10).json()
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices